Ensure the buffer provided to MPAS_io_get_var_generic is large enough.

jim-p-w opened the pull request:
This PR fixes a potential buffer overrun when reading string variables from a netcdf file.

A fixed size array is provided as an output buffer when reading a 0d-char character variable. 
Call MPAS_io_inq_var prior to the read to get the size of the variable, and only proceed with the read if the size of the variable will fit in the provided array.
Return an error code if the variable value is larger than the provided output buffer.

A unit test is included to verify:
1. An attempt to read a string variable into a buffer which is too small to hold the string value is detected
2. When a buffer which is too small is detected, the read won't occur and an error code is returned
3. When the test is run via valgrind, valgrind detects no memory errors or corruption
4. If the code which detects the too small buffer is commented out the test crashes and valgrind reports memory corruption

Fixes issue https://github.com/MPAS-Dev/MPAS-Model/issues/1350

### Note ###
When building with PIO, if the charArray (or the charArray1d) value exceeds the size of the tempchar buffer provided to the call to `PIO_get_var`, the value will be truncated to the size of the provided tempchar buffer (lines 2018, 2023, and 2042 in src/framework/mpas_io.F).

mgduda commented on src/core_test/Makefile:
```diff
@@ -12,7 +12,8 @@ OBJS = mpas_test_core.o \
        mpas_test_openacc.o \
-       mpas_test_core_stream_list.o
+       mpas_test_core_stream_list.o \
+       mpas_test_core_io.o \
```
Although it's not incorrect, for the sake of consistency and to avoid problems if someone were to inadvertently delete the blank line below, I'd suggest not using a trailing `\` for `mpas_test_core_io.o`.

jim-p-w replied:
I addressed this, plus added the new file to the `mpas_test_core.` dependencies.

mgduda commented on src/core_test/mpas_test_core.F:
```diff
@@ -224,6 +225,17 @@ function test_core_run(domain) result(iErr)!{{{
+      else
+          call mpas_log_write('$i tests FAILED', intArgs=[iErr])
+      end if
+      call mpas_log_write('')
```
I'd suggest adding a blank line below this one for consistency of style.

mgduda commented on src/core_test/mpas_test_core_io.F:
```diff
@@ -0,0 +1,208 @@
+! Copyright (c) 2013,  Los Alamos National Security, LLC (LANS)
```
At a minimum, we should update the year to 2025. But you can also follow the copyright statement in, e.g., [mpas_test_core_stream_list.F](https://github.com/MPAS-Dev/MPAS-Model/blob/develop/src/core_test/mpas_test_core_stream_list.F).

jim-p-w replied:
I used the copyright from `mpas_test_core_stream_list.F`

mgduda commented on src/core_test/mpas_test_core_io.F:
```diff
@@ -0,0 +1,208 @@
+
+      ! log an error message
+      if (present(message)) then
+        if (present(args)) then
```
Some places in this file use two spaces of indentation and others use three. I'd suggest sticking with a consistent indentation (with three or four spaces being the most common throughout MPAS, as far as I'm aware).

jim-p-w replied:
I think I've got everything uniformly using 3 spaces per indent level.

mgduda commented on src/core_test/mpas_test_core_io.F:
"strin" -> "string"

jim-p-w replied:
Fixed the typo.

mgduda commented on src/core_test/mpas_test_core_io.F:
We could replace `2` with `size(varNames)`.

jim-p-w replied:
Now using the `size()` function for look upper bound.

mgduda commented on src/core_test/mpas_test_core_io.F:
Similarly, we could use `size(varNames)` here rather than `2`.

mgduda commented on src/framework/mpas_io.F:
```diff
@@ -1984,22 +1993,40 @@ subroutine MPAS_io_get_var_generic(handle, fieldname, intVal, intArray1d, intArr
+         ! get the dimension of the char variable to ensure the provided output buffer is large enough
+         call MPAS_io_inq_var(handle, fieldname, fieldtype, ndims, dimnames, dimsizes, local_ierr)
+         do i = 1, ndims
+           message = '  MPAS_io_get_var_generic len(charVal):$i var "'//trim(fieldname)// &
```
I'd suggest following the surrounding code and using three spaces of indentation.

jim-p-w replied:
Think I've got everything using 3 spaces uniformly.

mgduda commented on src/framework/mpas_io.F:
```diff
@@ -6498,6 +6531,10 @@ subroutine MPAS_io_err_mesg(ioContext, ierr, fatal)
+         case (MPAS_IO_ERR_MISSING_DIM)
+            call mpas_log_write('MPAS IO Error: Attempting to read a dimension which does not exist.', MPAS_LOG_ERR)
+         case (MPAS_IO_ERR_INSUFFICIENT_ARG)
+            call mpas_log_write('MPAS IO Error: Attempting to read a string into a buffer which is too small.', MPAS_LOG_ERR)
```
Just a thought, but what if we named this error code `MPAS_IO_ERR_INSUFFICIENT_BUF`, and generalized the error string to something like: `MPAS IO Error: Attempting to read a variable into a buffer of insufficient size.` This might allow us to re-use this error condition more generally throughout the `mpas_io` module when we detect that there is too much data for the buffer that was provided, rather than having the error be specific to strings.

jim-p-w replied:
Good idea; implemented.

mgduda commented on src/framework/mpas_io.F:
```diff
@@ -1847,6 +1849,13 @@ subroutine MPAS_io_get_var_generic(handle, fieldname, intVal, intArray1d, intArr
+      integer :: ndims
+      integer, dimension(:), pointer :: dimsizes
+      character (len=StrKIND), dimension(:), pointer :: dimnames
+      character (len=StrKind) :: message
```
For consistency with the line above it and elsewhere in MPAS, let's change `StrKind` to `StrKIND` on this line.

jim-p-w replied:
Capitalization is amended.

mgduda commented on src/framework/mpas_io.F:
```diff
@@ -1984,22 +1993,40 @@ subroutine MPAS_io_get_var_generic(handle, fieldname, intVal, intArray1d, intArr
+            if (field_cursor % fieldhandle % has_unlimited_dim) then
+               count2(1) = field_cursor % fieldhandle % dims(1) % dimsize
+               pio_ierr = PIO_get_var(handle % pio_file, field_cursor % fieldhandle % fieldid, start2, count2, tempchar)
+              charVal(1:count2(1)) = tempchar(1)(1:count2(1))
```
I think one more space of indentation is needed here.

jim-p-w replied:
A vi search string of `/^\(   \)\+` (note 3 spaces between the parens) was quite helpful!

mgduda commented on src/framework/mpas_io.F:
```diff
@@ -1984,22 +1993,40 @@ subroutine MPAS_io_get_var_generic(handle, fieldname, intVal, intArray1d, intArr
+         ! because charVal is provided, assume dimension 1 is the string length
+         if (dimsizes(1) > len(charVal)) then
+            local_ierr = MPAS_IO_ERR_INSUFFICIENT_ARG
+            message = '  MPAS_io_get_var_generic var "'//trim(fieldname)//  &
```
Are the leading two spaces in this string intentional? They lead to messages that look like this:
```
ERROR:   MPAS_io_get_var_generic var "xtime" len too big, len=128 buflen=64
ERROR: MPAS IO Error: Attempting to read a string into a buffer which is too small.
```

jim-p-w replied:
whitespace is modified.

mgduda commented on src/framework/mpas_io.F:
```diff
@@ -2765,28 +2792,34 @@ subroutine MPAS_io_get_var_generic(handle, fieldname, intVal, intArray1d, intArr
+      if (local_ierr == MPAS_IO_ERR_INSUFFICIENT_BUF) then
+         call MPAS_io_err_mesg(handle % ioContext, local_ierr, .false.)
+         io_global_err = local_ierr
+         if (present(ierr)) ierr = local_ierr
```
In case of an `MPAS_IO_ERR_INSUFFICIENT_BUF` error, this will be the last statement executed in this subroutine, and there may be an opportunity to minimize the changes to the code below this block. Rather than logic that is structured like
```fortran
      if (local_ierr == MPAS_IO_ERR_INSUFFICIENT_ARG) then
         ... code ...
      else
#ifdef MPAS_PIO_SUPPORT
        if (pio_ierr /= PIO_noerr) then
           ... code ...
           return
        end if
#endif

#ifdef MPAS_SMIOL_SUPPORT
        if (local_ierr /= SMIOL_SUCCESS) then
           ... code ...
           return
        end if
#endif
      end if
```
what if we used something like this:
```fortran
      if (local_ierr == MPAS_IO_ERR_INSUFFICIENT_ARG) then
         ... code ...
         return
      end if

#ifdef MPAS_PIO_SUPPORT
      if (pio_ierr /= PIO_noerr) then
         ... code ...
         return
      end if
#endif

#ifdef MPAS_SMIOL_SUPPORT
      if (local_ierr /= SMIOL_SUCCESS) then
         ... code ...
         return
      end if
#endif
```
The idea here is that everything from the `#ifdef MPAS_PIO_SUPPORT` line downward is completely unchanged from the base code.

jim-p-w replied:
Nice observation. Done.

mgduda commented on src/framework/mpas_io.F:
```diff
@@ -7,6 +7,8 @@
 module mpas_io
 
+#define IO_DEBUG_WRITE(M, ARGS) !call mpas_log_write(M, ARGS)
+#define IO_ERROR_WRITE(M, ARGS) call mpas_log_write( M, ARGS, messageType=MPAS_LOG_ERR)
```
So that we don't need to make up an unused argument in cases where we don't have any arguments to the `mpas_log_write` call, for example:
```
IO_ERROR_WRITE('Here is a message', intArgs=[0])
```
we could instead define this macro as:
```
#define IO_ERROR_WRITE(M) call mpas_log_write(M, messageType=MPAS_LOG_ERR)
```
In order to allow for the passing of other arguments to `mpas_log_write`, we can also define the macro `COMMA` as:
```
#define COMMA ,
```
Then, an error message with arguments would look something like:
```
IO_ERROR_WRITE('The answer is $i' COMMA intArgs=[42])
```

jim-p-w replied:
Clever! I had wondered about the definition of `COMMA` in other files. Done.

mgduda commented on src/framework/mpas_io.F:
```diff
@@ -1984,22 +1993,40 @@ subroutine MPAS_io_get_var_generic(handle, fieldname, intVal, intArray1d, intArr
+         if (dimsizes(1) > len(charVal)) then
+            local_ierr = MPAS_IO_ERR_INSUFFICIENT_BUF
+            message = 'MPAS_io_get_var_generic var "'//trim(fieldname)//  &
+                      '" len too big, len=$i buflen=$i'
```
It might make this message more helpful to users if we used something like:
```fortran
            message = 'Length of string variable "'//trim(fieldname)//'" in file "'//trim(handle % filename)//'"'
            IO_ERROR_WRITE(message, intArgs=(/0/))
            message = '    exceeds buffer size: len('//trim(fieldname)//')=$i, len(buffer)=$i'
            IO_ERROR_WRITE(message, intArgs=(/dimsizes(1), len(charVal)/)) 
```
This would generate an error log entry something like:
```
ERROR: Length of string variable "xtime" in file "x1.10242.init.nc"
ERROR:     exceeds buffer size: len(xtime)=128, len(buffer)=64
```

jim-p-w replied:
Good idea. I love printing out the filename which contains the variable. Done.

mgduda commented on src/framework/mpas_io.F:
```diff
@@ -2765,6 +2794,12 @@ subroutine MPAS_io_get_var_generic(handle, fieldname, intVal, intArray1d, intArr
+         io_global_err = local_ierr
+         if (present(ierr)) ierr = local_ierr
+         return
+      endif
```
A blank line below this `endif` might be helpful in providing a visual indicator that the block below is independent of this block that is being terminated here.

jim-p-w replied:
Blank line has been added.

mgduda commented on src/core_test/mpas_test_core_io.F:
With the GCC compilers and using the `-Wall` flag, there's a warning that `threadErrs` is never set:
```
mpas_test_core_io.F:61:62:

   61 |    subroutine test_read_string_buffer_check(domain, threadErrs, ierr)
      |                                                              1
Warning: Dummy argument 'threaderrs' at (1) was declared INTENT(OUT) but was not set [-Wunused-dummy-argument]
```
We could either ensure that this `intent(out)` argument is set, or -- since it doesn't appear to be used by any calling code -- we could just eliminate this argument.

jim-p-w replied:
threaderrs argument has been removed from test_core_io_test and test_read_string_buffer_check

mgduda commented on src/framework/mpas_io.F:
```diff
@@ -1984,22 +1994,41 @@ subroutine MPAS_io_get_var_generic(handle, fieldname, intVal, intArray1d, intArr
+            message = 'Length of string variable "'//trim(fieldname)//'" in file "'//trim(handle % filename)//'"'
+            IO_ERROR_WRITE(message)
+            message = '    exceeds buffer size: len('//trim(fieldname)//')=$i, len(buffer)=$i'
+            IO_ERROR_WRITE(message COMMA intArgs=(/dimsizes(1), len(charVal)/))
```
With the GCC 12.4.0 compilers, this macro seems to work as expected, and I get a message like
```
ERROR: Length of string variable "xtime" in file "x1.10242.init.nc"
ERROR:     exceeds buffer size: len(xtime)=128, len(buffer)=64
```
in the log file for a test that I've been using.

However, with the Intel oneAPI 2025.1.0 compilers, I get
```
ERROR: Length of string variable "xtime" in file "x1.10242.init.nc"
ERROR:     exceeds buffer size: len(xtime)=**, len(buffer)=**
```

The problem seems to be in the macro expansion with the `ifx` compiler. Using the `-preprocess-only` option, it appears the `ifx` compiler is expanding the macro as:
```fortran
            call mpas_log_write( message, messageType=MPAS_LOG_ERR)
```

Could you investigate whether there's a workaround that we can employ so that this works with the `ifx` compiler?

jim-p-w replied:
I have changed the log macros to use two explicit arguments while I investigate a one argument solution which works with the intel compiler.

mgduda commented on src/core_test/mpas_test_core_io.F:
```diff
@@ -0,0 +1,203 @@
+
+      ! log an error message
+      if (present(message)) then
+         if (present(args)) then
```
I think optional arguments can be passed through to other subroutines whose corresponding dummy arguments are also optional without any need to check presence. So it may be possible to simplify this block of code as:
```fortran
      if (present(message)) then
         ERROR_WRITE_ARGS(message, intArgs=args)
      end if
```

jim-p-w replied:
I see.

mgduda commented on src/core_test/mpas_test_core_io.F:
Although it apparently does no harm, are the two spaces in the `'Time  '` dimension intentional?

mgduda replied:
Ok, I see there's a compilation error if these strings don't have the same length:
```
mpas_test_core_io.F:65:84:

   65 |       character (len=StrKIND), dimension(2), parameter :: dimNamesStringTime = ['StrLen', 'Time']
      |                                                                                    1
Error: Different CHARACTER lengths (6/4) in array constructor at (1)
```

It should work to declare `dimNamesStringTime` as:
```fortran
character (len=StrKIND), dimension(2), parameter :: dimNamesStringTime = [character(len=StrKIND) :: 'StrLen', 'Time']
```

jim-p-w replied:
Thanks, I wouldn't have figured that out!

jim-p-w replied:
Thanks, I wouldn't have figured that out!

mgduda commented on src/core_test/mpas_test_core_io.F:
Out of curiosity, is there a reason for specifying a length of `32` for these strings and a length of `StrKIND` for the strings on lines 68 - 69?

mgduda commented on src/core_test/mpas_test_core_io.F:
Since we're not writing any large variables (and so parallel I/O isn't essential), it might increase compatibility if we used `MPAS_IO_NETCDF` for the format here rather than `MPAS_IO_NETCDF4`.

jim-p-w replied:
OK, but that didn't change the file type. ncdump -k reports `cdf5` for both filetypes of `MPAS_IO_NETCDF` and `MPAS_IO_NETCDF4`

jim-p-w replied:
OK. It looks like that parameter isn't used with SMIOL? ncdump -k reports `cdf5` when using `MPAS_IO_NETCDF4` and when using `MPAS_IO_NETCDF4`.

mgduda replied:
Right -- SMIOL ignores `ioformat` at present, but the PIO library does make use of `ioformat` when creating new files.

mgduda reviewed (approved):
@jim-p-w I've verified that the changes to `mpas_io.F` catch cases where we would otherwise have tried to read a string into a buffer that was too small. If you could rework the commit history so that it contains just a couple of commits -- perhaps one to modify the framework mpas_io code, and another to add unit tests -- I think we'll be ready to merge this.

jim-p-w commented:
@mgduda Is the initial commit message in the PR suitable for the framework commit? Or do you have suggestions/edits you would prefer to see?

jim-p-w committed:
Ensure the buffer provided to MPAS_io_get_var_generic is large enough.

A fixed size array is provided as an input buffer when reading 0d-char
character variables.
Call MPAS_io_inq_var prior to the read to get the size of the variable’s value,
and only proceed with the read if the variable’s value will fit in the
provided array.
Return an error code if the variable’s value is larger than the provided
input buffer.

jim-p-w committed:
Add a test to verify reading character variables won't overrun buffers.

Character variables are read into fixed size arrays when reading netcdf
files. A test is added which tries to read character variables into a
buffer which isn't large enough to hold the data. The test verifies the
read fails with a suitable error code.

jim-p-w committed:
Allow optional arguments to logging macros in mpas_io.F

IO_DEBUG_WRITE and IO_ERROR_WRITE take one argument,
which is passed to mpas_log_write().
To pass parameters to mpas_log_write invoke the macros
providing COMMA between arguments, e.g.
IO_DEBUG_WRITE('message' COMMA intArgs=[42])

mgduda commented on src/core_test/mpas_test_core_io.F:
```diff
@@ -0,0 +1,200 @@
+module test_core_io
+
+#define ERROR_WRITE(M) call mpas_log_write( M , messageType=MPAS_LOG_ERR)
+#define ERROR_WRITE_ARGS(M, ARGS) call mpas_log_write( M , ARGS, messageType=MPAS_LOG_ERR)
```
Rather than two macros here for writing error messages, I suspect we could use the same `COMMA` macro approach as in `mpas_io.F` and have just a single `ERROR_WRITE` macro.

mgduda replied:
On second thought, since the code is correct as written, let's just keep the macros as they are. Better to get this PR merged, in my opinion!
